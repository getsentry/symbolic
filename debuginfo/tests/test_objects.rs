use std::fmt;

use failure::Error;
use insta;

use symbolic_common::ByteView;
use symbolic_debuginfo::{DebugSession, Function, Object, SymbolMap};

/// Helper to create neat snapshots for symbol tables.
struct SymbolsDebug<'a>(&'a SymbolMap<'a>);

impl fmt::Debug for SymbolsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for symbol in self.0.iter() {
            writeln!(
                f,
                "{:>16x} {}",
                &symbol.address,
                &symbol.name().unwrap_or("<unknown>")
            )?;
        }

        Ok(())
    }
}

/// Helper to create neat snapshots for function trees.
struct FunctionsDebug<'a>(&'a [Function<'a>], usize);

impl fmt::Debug for FunctionsDebug<'_> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        for function in self.0 {
            writeln!(
                f,
                "\n{:indent$}> {:#x}: {} ({:#x})",
                "",
                function.address,
                function.name,
                function.size,
                indent = self.1 * 2
            )?;

            for line in &function.lines {
                writeln!(
                    f,
                    "{:indent$}  {:#x}: {}:{} ({})",
                    "",
                    line.address,
                    line.file.name_str(),
                    line.line,
                    line.file.dir_str(),
                    indent = self.1 * 2
                )?;
            }

            write!(f, "{:?}", FunctionsDebug(&function.inlinees, self.1 + 1))?;
        }

        Ok(())
    }
}

#[test]
fn test_breakpad() -> Result<(), Error> {
    // Using the windows version here since it contains all record kinds
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Breakpad(
    BreakpadObject {
        code_id: Some(
            CodeId(5ab380779000)
        ),
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        arch: X86,
        name: "crash.pdb",
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_breakpad_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("breakpad_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_breakpad_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.sym")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!(
        "breakpad_functions",
        FunctionsDebug(&functions[..10], 0)
    );

    Ok(())
}

#[test]
fn test_elf_executable() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Elf(
    ElfObject {
        code_id: Some(
            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c)
        ),
        debug_id: DebugId {
            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
            appendix: 0
        },
        arch: Amd64,
        kind: Executable,
        load_address: 0x400000,
        has_symbols: true,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_elf_debug() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash.debug")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Elf(
    ElfObject {
        code_id: Some(
            CodeId(f1c3bcc0279865fe3058404b2831d9e64135386c)
        ),
        debug_id: DebugId {
            uuid: "c0bcc3f1-9827-fe65-3058-404b2831d9e6",
            appendix: 0
        },
        arch: Amd64,
        kind: Debug,
        load_address: 0x400000,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_elf_symbols() -> Result<(), Error> {
    // TODO(ja): Why does crash.debug not retain the symbol table but report has_symbols
    let view = ByteView::open("../testutils/fixtures/linux/crash")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("elf_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_elf_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/linux/crash.debug")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("elf_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_mach_executable() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"MachO(
    MachObject {
        code_id: Some(
            CodeId(67e9247c814e392ba027dbde6748fcbf)
        ),
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        arch: Amd64,
        kind: Executable,
        load_address: 0x100000000,
        has_symbols: true,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_mach_dsym() -> Result<(), Error> {
    let view =
        ByteView::open("../testutils/fixtures/macos/crash.dSYM/Contents/Resources/DWARF/crash")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"MachO(
    MachObject {
        code_id: Some(
            CodeId(67e9247c814e392ba027dbde6748fcbf)
        ),
        debug_id: DebugId {
            uuid: "67e9247c-814e-392b-a027-dbde6748fcbf",
            appendix: 0
        },
        arch: Amd64,
        kind: Debug,
        load_address: 0x100000000,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_mach_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/macos/crash")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("mach_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_mach_functions() -> Result<(), Error> {
    let view =
        ByteView::open("../testutils/fixtures/macos/crash.dSYM/Contents/Resources/DWARF/crash")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("mach_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}

#[test]
fn test_pe_32() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.exe")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pe(
    PeObject {
        code_id: Some(
            CodeId(5ab380779000)
        ),
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        debug_file_name: Some(
            "C:\\projects\\breakpad-tools\\windows\\Release\\crash.pdb"
        ),
        arch: X86,
        kind: Executable,
        load_address: 0x400000,
        has_symbols: false,
        has_debug_info: false,
        has_unwind_info: false
    }
)"###);

    Ok(())
}

#[test]
fn test_pe_64() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/CrashWithException.exe")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pe(
    PeObject {
        code_id: Some(
            CodeId(5c9e09599000)
        ),
        debug_id: DebugId {
            uuid: "f535c5fb-2ae8-4bb8-aa20-6c30be566c5a",
            appendix: 1
        },
        debug_file_name: Some(
            "C:\\Users\\sentry\\source\\repos\\CrashWithException\\x64\\Release\\CrashWithException.pdb"
        ),
        arch: Amd64,
        kind: Executable,
        load_address: 0x140000000,
        has_symbols: false,
        has_debug_info: false,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

// NB: No test for PE symbols because our executable does not export any symbols
// NB: No test for PE functions because we can only read debug info from PDBs

#[test]
fn test_pdb() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    insta::assert_debug_snapshot_matches!(object, @r###"Pdb(
    PdbObject {
        debug_id: DebugId {
            uuid: "3249d99d-0c40-4931-8610-f4e4fb0b6936",
            appendix: 1
        },
        arch: X86,
        load_address: 0x0,
        has_symbols: true,
        has_debug_info: true,
        has_unwind_info: true
    }
)"###);

    Ok(())
}

#[test]
fn test_pdb_symbols() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    let symbols = object.symbol_map();
    insta::assert_debug_snapshot_matches!("pdb_symbols", SymbolsDebug(&symbols));

    Ok(())
}

#[test]
fn test_pdb_functions() -> Result<(), Error> {
    let view = ByteView::open("../testutils/fixtures/windows/crash.pdb")?;
    let object = Object::parse(&view)?;

    let mut session = object.debug_session()?;
    let functions = session.functions().collect::<Result<Vec<_>, _>>()?;
    insta::assert_debug_snapshot_matches!("pdb_functions", FunctionsDebug(&functions[..10], 0));

    Ok(())
}
