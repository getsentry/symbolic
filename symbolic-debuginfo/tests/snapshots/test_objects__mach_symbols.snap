---
created: "2019-02-20T12:28:37.426989Z"
creator: insta@0.6.3
source: debuginfo/tests/test_objects.rs
expression: SymbolDebug(&symbols)
---
               0 _mh_execute_header
             d20 _ZN15google_breakpad18MinidumpFileWriterC2Ev
             d40 _ZN15google_breakpad18MinidumpFileWriterC1Ev
             d60 _ZN15google_breakpad18MinidumpFileWriterD2Ev
             da0 _ZN15google_breakpad18MinidumpFileWriter5CloseEv
             de0 _ZN15google_breakpad18MinidumpFileWriterD1Ev
             e20 _ZN15google_breakpad18MinidumpFileWriter4OpenEPKc
             e70 _ZN15google_breakpad18MinidumpFileWriter7SetFileEi
             ea0 _ZN15google_breakpad18MinidumpFileWriter20CopyStringToMDStringEPKwjPNS_10TypedMDRVAI8MDStringEE
             ff0 _ZN15google_breakpad18MinidumpFileWriter20CopyStringToMDStringEPKcjPNS_10TypedMDRVAI8MDStringEE
            1140 _ZN15google_breakpad18MinidumpFileWriter11WriteStringEPKwjP20MDLocationDescriptor
            1150 _ZN15google_breakpad18MinidumpFileWriter15WriteStringCoreIwEEbPKT_jP20MDLocationDescriptor
            1420 _ZN15google_breakpad18MinidumpFileWriter11WriteStringEPKcjP20MDLocationDescriptor
            1430 _ZN15google_breakpad18MinidumpFileWriter15WriteStringCoreIcEEbPKT_jP20MDLocationDescriptor
            1700 _ZN15google_breakpad18MinidumpFileWriter11WriteMemoryEPKvmP18MDMemoryDescriptor
            17d0 _ZN15google_breakpad12UntypedMDRVA8AllocateEm
            18e0 _ZN15google_breakpad18MinidumpFileWriter8AllocateEm
            19a0 _ZN15google_breakpad18MinidumpFileWriter4CopyEjPKvl
            1a70 _ZN15google_breakpad12UntypedMDRVA4CopyEjPKvm
            1b70 _ZN15google_breakpad21CrashGenerationClient23RequestDumpForExceptionEiiij
            1d30 _ZN15google_breakpad14ReadTaskMemoryEjymRNSt3__16vectorIhNS0_9allocatorIhEEEE
            1df0 _ZN15google_breakpad12DynamicImage29CalculateMemoryAndVersionInfoEv
            1fb0 _ZN15google_breakpad12DynamicImage11GetFileTypeEv
            1fc0 _ZN15google_breakpad13DynamicImagesC2Ej
            2080 _ZN15google_breakpad13DynamicImages20DetermineTaskCPUTypeEj
            2160 _ZN15google_breakpad13DynamicImages20ReadImageInfoForTaskEv
            21c0 _ZN15google_breakpad13DynamicImagesC1Ej
            21d0 _ZN15google_breakpad13DynamicImages27GetDyldAllImageInfosPointerEv
            2210 _ZN15google_breakpad13ReadImageInfoINS_7MachO64EEEvRNS_13DynamicImagesEy
            27e0 _ZN15google_breakpad13ReadImageInfoINS_7MachO32EEEvRNS_13DynamicImagesEy
            2db0 _ZN15google_breakpad13DynamicImages18GetExecutableImageEv
            2e00 _ZN15google_breakpad13DynamicImages23GetExecutableImageIndexEv
            2e40 _ZNSt3__16vectorIhNS_9allocatorIhEEE8__appendEm
            3010 _ZN15google_breakpadL14ReadTaskStringEjy
            31d0 _ZN15google_breakpad12DynamicImageC2EPhmyNSt3__112basic_stringIcNS2_11char_traitsIcEENS2_9allocatorIcEEEEmji
            32f0 _ZNSt3__16vectorIN15google_breakpad15DynamicImageRefENS_9allocatorIS2_EEE21__push_back_slow_pathIS2_EEvOT_
            3480 _ZNSt3__16__sortIRNS_6__lessIN15google_breakpad15DynamicImageRefES3_EEPS3_EEvT0_S7_T_
            3aa0 _ZNSt3__17__sort5IRNS_6__lessIN15google_breakpad15DynamicImageRefES3_EEPS3_EEjT0_S7_S7_S7_S7_T_
            3bf0 _ZNSt3__127__insertion_sort_incompleteIRNS_6__lessIN15google_breakpad15DynamicImageRefES3_EEPS3_EEbT0_S7_T_
            3ec0 _ZN15google_breakpad19breakpad_exc_serverEP17mach_msg_header_tS1_
            3ed0 catch_exception_raise
            3ef0 _ZN15google_breakpad16ForwardExceptionEjjiPij
            3ff0 _ZN15google_breakpad16ExceptionHandlerC2ERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEPFbPvEPFbPKcSE_SA_bESA_bSE_
            4180 _ZN15google_breakpad16ExceptionHandler5SetupEb
            42a0 _ZN15google_breakpad10scoped_ptrINS_21CrashGenerationClientEED1Ev
            42b0 _ZN15google_breakpad10scoped_ptrI9sigactionED1Ev
            42c0 _ZN15google_breakpad16ExceptionHandlerC1ERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEPFbPvEPFbPKcSE_SA_bESA_bSE_
            42d0 _ZN15google_breakpad16ExceptionHandlerC2EPFbPviiijES1_b
            4400 _ZN15google_breakpad16ExceptionHandlerC1EPFbPviiijES1_b
            4410 _ZN15google_breakpad16ExceptionHandlerD2Ev
            4580 _ZN15google_breakpad16ExceptionHandler8TeardownEv
            4680 __clang_call_terminate
            4690 _ZN15google_breakpad16ExceptionHandlerD1Ev
            46a0 _ZN15google_breakpad16ExceptionHandler13WriteMinidumpEb
            47c0 _ZN15google_breakpad16ExceptionHandler26SendMessageToHandlerThreadENS_20HandlerThreadMessageE
            4880 _ZN15google_breakpad16ExceptionHandler12UpdateNextIDEv
            4940 _ZN15google_breakpad16ExceptionHandler13WriteMinidumpERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEbPFbPKcSB_PvbESC_
            49e0 _ZN15google_breakpad16ExceptionHandler21WriteMinidumpForChildEjjRKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEPFbPKcSB_PvbESC_
            4b20 _ZN15google_breakpad16ExceptionHandler26WriteMinidumpWithExceptionEiiiP17__darwin_ucontextjbb
            4d80 _ZN15google_breakpad16ExceptionHandler14WaitForMessageEPv
            5060 _ZN15google_breakpad16ExceptionHandler14SuspendThreadsEv
            50d0 _ZN15google_breakpad16ExceptionHandler13ResumeThreadsEv
            5140 _ZN15google_breakpad16ExceptionHandler16UninstallHandlerEb
            5230 _ZN15google_breakpad16ExceptionHandler13SignalHandlerEiP9__siginfoPv
            5280 _ZN15google_breakpad16ExceptionHandler14InstallHandlerEv
            53f0 _ZN15google_breakpad17MinidumpGeneratorC2Ev
            5710 _ZN15google_breakpad17MinidumpGenerator23GatherSystemInformationEv
            5970 _ZN15google_breakpad17MinidumpGeneratorC1Ev
            5980 _ZN15google_breakpad17MinidumpGeneratorC2Ejj
            5cd0 _ZN15google_breakpad17MinidumpGeneratorC1Ejj
            5ce0 _ZN15google_breakpad17MinidumpGeneratorD2Ev
            5d80 _ZN15google_breakpad17MinidumpGeneratorD1Ev
            5d90 _ZN15google_breakpad17MinidumpGeneratorD0Ev
            5db0 _ZN15google_breakpad17MinidumpGenerator14SetTaskContextEP17__darwin_ucontext
            5dc0 _ZN15google_breakpad17MinidumpGenerator21UniqueNameInDirectoryERKNSt3__112basic_stringIcNS1_11char_traitsIcEENS1_9allocatorIcEEEEPS7_
            5ee0 _ZN15google_breakpad17MinidumpGenerator5WriteEPKc
            61a0 _ZN15google_breakpad17MinidumpGenerator21WriteThreadListStreamEP14MDRawDirectory
            63f0 _ZN15google_breakpad17MinidumpGenerator21WriteMemoryListStreamEP14MDRawDirectory
            6910 _ZN15google_breakpad17MinidumpGenerator21WriteSystemInfoStreamEP14MDRawDirectory
            6b70 _ZN15google_breakpad17MinidumpGenerator21WriteModuleListStreamEP14MDRawDirectory
            6e00 _ZN15google_breakpad17MinidumpGenerator19WriteMiscInfoStreamEP14MDRawDirectory
            7060 _ZN15google_breakpad17MinidumpGenerator23WriteBreakpadInfoStreamEP14MDRawDirectory
            7170 _ZN15google_breakpad17MinidumpGenerator18CalculateStackSizeEy
            7290 _ZN15google_breakpad17MinidumpGenerator26WriteStackFromStartAddressEyP18MDMemoryDescriptor
            7410 _ZN15google_breakpad17MinidumpGenerator10WriteStackEPjP18MDMemoryDescriptor
            7440 _ZN15google_breakpad17MinidumpGenerator13WriteStackX86EPjP18MDMemoryDescriptor
            7450 _ZN15google_breakpad17MinidumpGenerator16WriteStackX86_64EPjP18MDMemoryDescriptor
            7460 _ZN15google_breakpad17MinidumpGenerator12WriteContextEPjP20MDLocationDescriptor
            7490 _ZN15google_breakpad17MinidumpGenerator15WriteContextX86EPjP20MDLocationDescriptor
            7640 _ZN15google_breakpad17MinidumpGenerator18WriteContextX86_64EPjP20MDLocationDescriptor
            7810 _ZN15google_breakpad17MinidumpGenerator17CurrentPCForStackEPj
            7850 _ZN15google_breakpad17MinidumpGenerator20CurrentPCForStackX86EPj
            7860 _ZN15google_breakpad17MinidumpGenerator23CurrentPCForStackX86_64EPj
            7870 _ZN15google_breakpad17MinidumpGenerator14GetThreadStateEjPjS1_
            7930 _ZN15google_breakpad17MinidumpGenerator17WriteThreadStreamEjP11MDRawThread
            7ae0 _ZN15google_breakpad17MinidumpGenerator20WriteExceptionStreamEP14MDRawDirectory
            7d60 _ZN15google_breakpad17MinidumpGenerator17WriteModuleStreamEjP11MDRawModule
            8100 _ZN15google_breakpad17MinidumpGenerator20FindExecutableModuleEv
            8150 _ZN15google_breakpad17MinidumpGenerator13WriteCVRecordEP11MDRawModuleiPKcb
            8500 _ZN15google_breakpad13DynamicImagesD2Ev
            85b0 _ZNSt3__114__split_bufferI18MDMemoryDescriptorRN15google_breakpad16PageStdAllocatorIS1_EEEC2EmmS5_
            8710 _ZNSt3__16vectorI18MDMemoryDescriptorN15google_breakpad16PageStdAllocatorIS1_EEE21__push_back_slow_pathIRKS1_EEvOT_
            8990 ConvertUTF32toUTF16
            8b40 ConvertUTF16toUTF32
            8c60 ConvertUTF16toUTF8
            8f70 isLegalUTF8Sequence
            90d0 ConvertUTF8toUTF16
            93c0 ConvertUTF32toUTF8
            9630 ConvertUTF8toUTF32
            9b20 _ZN15google_breakpad7MD5InitEPNS_10MD5ContextE
            9b40 _ZN15google_breakpad9MD5UpdateEPNS_10MD5ContextEPKhm
            9c80 _ZN15google_breakpadL12MD5TransformEPjPKj
            a370 _ZN15google_breakpad8MD5FinalEPhPNS_10MD5ContextE
            a480 _ZN15google_breakpad11UTF8ToUTF16EPKcPNSt3__16vectorItNS2_9allocatorItEEEE
            a570 _ZNSt3__16vectorItNS_9allocatorItEEE6insertENS_11__wrap_iterIPKtEEmRS5_
            ac00 _ZN15google_breakpad15UTF8ToUTF16CharEPKciPt
            ac80 _ZN15google_breakpad12UTF32ToUTF16EPKwPNSt3__16vectorItNS2_9allocatorItEEEE
            ad70 _ZN15google_breakpad16UTF32ToUTF16CharEwPt
            adc0 _ZN15google_breakpad11UTF16ToUTF8ERKNSt3__16vectorItNS0_9allocatorItEEEEb
            b080 _ZNSt3__16vectorItNS_9allocatorItEEE8__appendEm
            b1a0 _ZN8breakpad17BootstrapRegisterEjPcj
            b1b0 _ZN15google_breakpad6FileIDC2EPKc
            b1d0 _ZN15google_breakpad6FileIDC1EPKc
            b1f0 _ZN15google_breakpad6FileID14FileIdentifierEPh
            b2d0 _ZN15google_breakpad6FileID15MachoIdentifierEiiPh
            b380 _ZN15google_breakpad6FileID25ConvertIdentifierToStringEPKhPci
            b410 _ZN16MacFileUtilities7MachoIDC2EPKc
            b4e0 _ZN16MacFileUtilities7MachoIDC1EPKc
            b5b0 _ZN16MacFileUtilities7MachoIDC2EPKcPvm
            b670 _ZN16MacFileUtilities7MachoIDC1EPKcPvm
            b730 _ZN16MacFileUtilities7MachoIDD2Ev
            b740 _ZN16MacFileUtilities7MachoIDD1Ev
            b750 _ZN16MacFileUtilities7MachoID9UpdateCRCEPhm
            b9d0 _ZN16MacFileUtilities7MachoID9UpdateMD5EPhm
            b9e0 _ZN16MacFileUtilities7MachoID6UpdateEPNS_11MachoWalkerExm
            bad0 _ZN16MacFileUtilities7MachoID11UUIDCommandEiiPh
            bb40 _ZN16MacFileUtilities7MachoID10WalkHeaderEiiPFbPNS_11MachoWalkerEP12load_commandxbPvES5_
            bbd0 _ZN16MacFileUtilities7MachoID12UUIDWalkerCBEPNS_11MachoWalkerEP12load_commandxbPv
            bc10 _ZN16MacFileUtilities7MachoID9IDCommandEiiPh
            bd00 _ZN16MacFileUtilities7MachoID10IDWalkerCBEPNS_11MachoWalkerEP12load_commandxbPv
            bd40 _ZN16MacFileUtilities7MachoID7Adler32Eii
            bd80 _ZN16MacFileUtilities7MachoID8WalkerCBEPNS_11MachoWalkerEP12load_commandxbPv
            c320 _ZN16MacFileUtilities7MachoID3MD5EiiPh
            c390 _Z26breakpad_swap_uuid_commandP21breakpad_uuid_command
            c3a0 _Z26breakpad_swap_load_commandP12load_command
            c3b0 _Z27breakpad_swap_dylib_commandP13dylib_command
            c3e0 _Z29breakpad_swap_segment_commandP15segment_command
            c420 _Z32breakpad_swap_segment_command_64P18segment_command_64
            c490 _Z24breakpad_swap_fat_headerP10fat_header
            c4a0 _Z22breakpad_swap_fat_archP8fat_archj
            c530 _Z25breakpad_swap_mach_headerP11mach_header
            c560 _Z28breakpad_swap_mach_header_64P14mach_header_64
            c590 _Z21breakpad_swap_sectionP7sectionj
            c660 _Z24breakpad_swap_section_64P10section_64j
            c6c0 _ZN16MacFileUtilities11MachoWalkerC2EPKcPFbPS0_P12load_commandxbPvES6_
            c710 _ZN16MacFileUtilities11MachoWalkerC1EPKcPFbPS0_P12load_commandxbPvES6_
            c760 _ZN16MacFileUtilities11MachoWalkerC2EPvmPFbPS0_P12load_commandxbS1_ES1_
            c790 _ZN16MacFileUtilities11MachoWalkerC1EPvmPFbPS0_P12load_commandxbS1_ES1_
            c7c0 _ZN16MacFileUtilities11MachoWalkerD2Ev
            c7e0 _ZN16MacFileUtilities11MachoWalkerD1Ev
            c800 _ZN16MacFileUtilities11MachoWalker10WalkHeaderEii
            c960 _ZN16MacFileUtilities11MachoWalker10FindHeaderEiiRx
            cd00 _ZN16MacFileUtilities11MachoWalker20WalkHeader64AtOffsetEx
            cdf0 _ZN16MacFileUtilities11MachoWalker18WalkHeaderAtOffsetEx
            cf10 _ZN16MacFileUtilities11MachoWalker9ReadBytesEPvmx
            cf70 _ZN16MacFileUtilities11MachoWalker13CurrentHeaderEP14mach_header_64Px
            cfb0 _ZN16MacFileUtilities11MachoWalker14WalkHeaderCoreExjb
            d160 _ZN14MacStringUtils15ConvertToStringEPK10__CFString
            d270 _ZN14MacStringUtils19IntegerValueAtIndexERNSt3__112basic_stringIcNS0_11char_traitsIcEENS0_9allocatorIcEEEEj
            d590 _ZN15google_breakpad15MachSendMessageC2Ei
            d5e0 _ZN15google_breakpad11MachMessage18SetDescriptorCountEi
            d600 _ZN15google_breakpad11MachMessage7SetDataEPvi
            d680 _ZN15google_breakpad15MachSendMessageC1Ei
            d6c0 _ZN15google_breakpad11MachMessage13CalculateSizeEv
            d6e0 _ZN15google_breakpad11MachMessage13GetDataPacketEv
            d6f0 _ZN15google_breakpad11MachMessage13SetDescriptorEiRKNS_21MachMsgPortDescriptorE
            d710 _ZN15google_breakpad11MachMessage13AddDescriptorERKNS_21MachMsgPortDescriptorE
            d7c0 _ZN15google_breakpad11MachMessage13GetDescriptorEi
            d7e0 _ZN15google_breakpad11MachMessage17GetTranslatedPortEi
            d800 _ZN15google_breakpad11ReceivePortC2EPKc
            d890 _ZN15google_breakpad11ReceivePortC1EPKc
            d920 _ZN15google_breakpad11ReceivePortC2Ev
            d960 _ZN15google_breakpad11ReceivePortC1Ev
            d9a0 _ZN15google_breakpad11ReceivePortC2Ej
            d9b0 _ZN15google_breakpad11ReceivePortC1Ej
            d9c0 _ZN15google_breakpad11ReceivePortD2Ev
            d9f0 _ZN15google_breakpad11ReceivePortD1Ev
            da20 _ZN15google_breakpad11ReceivePort14WaitForMessageEPNS_18MachReceiveMessageEj
            da90 _ZN15google_breakpad14MachPortSenderC2EPKc
            dae0 _ZN15google_breakpad14MachPortSenderC1EPKc
            db30 _ZN15google_breakpad14MachPortSenderC2Ej
            db40 _ZN15google_breakpad14MachPortSenderC1Ej
            db50 _ZN15google_breakpad14MachPortSender11SendMessageERNS_15MachSendMessageEj
            dba0 main
            dc70 _ZN12_GLOBAL__N_18callbackEPKcS1_Pvb

